src/win_handler.rs

```diff
@@ -0,0 +1,458 @@
+///
+/// This is something of an internal detail and possibly we don't want to surface
+/// it publicly.
+pub struct DruidHandler<T: Data> {
+    /// The shared app state
+    app_state: Rc<RefCell<AppState<T>>>,
+    /// The id for the currenet window.
+    window_id: WindowId,
+}
+
+/// State shared by all windows in the UI.
+pub(crate) struct AppState<T: Data> {
+    command_queue: VecDeque<(WindowId, Command)>,
+    windows: Windows<T>,
+    pub(crate) env: Env,
+    pub(crate) data: T,
+}
+
+/// All active windows.
+struct Windows<T: Data> {
+    windows: HashMap<WindowId, Window<T>>,
+    state: HashMap<WindowId, WindowState>,
+}
+
+/// Per-window state not owned by user code.
+pub(crate) struct WindowState {
+    pub(crate) handle: WindowHandle,
+    prev_paint_time: Option<Instant>,
+}
+
+/// Everything required for a window to handle an event.
+struct SingleWindowState<'a, T: Data> {
+    window_id: WindowId,
+    window: &'a mut Window<T>,
+    state: &'a mut WindowState,
+    command_queue: &'a mut VecDeque<(WindowId, Command)>,
+    data: &'a mut T,
+    env: &'a Env,
+}
+
+impl<T: Data> Windows<T> {
+    fn connect(&mut self, id: WindowId, handle: WindowHandle) {
+        let state = WindowState {
+            handle,
+            prev_paint_time: None,
+        };
+        self.state.insert(id, state);
+    }
+
+    fn add(&mut self, id: WindowId, window: Window<T>) {
+        self.windows.insert(id, window);
+    }
+
+    fn remove(&mut self, id: WindowId) -> Option<WindowHandle> {
+        self.windows.remove(&id);
+        self.state.remove(&id).map(|state| state.handle)
+    }
+
+    //TODO: rename me?
+    fn get<'a>(
+        &'a mut self,
+        window_id: WindowId,
+        command_queue: &'a mut VecDeque<(WindowId, Command)>,
+        data: &'a mut T,
+        env: &'a Env,
+    ) -> Option<SingleWindowState<'a, T>> {
+        let state = self.state.get_mut(&window_id);
+        let window = self.windows.get_mut(&window_id);
+
+        match (state, window) {
+            (Some(state), Some(window)) => {
+                return Some(SingleWindowState {
+                    window_id,
+                    window,
+                    state,
+                    command_queue,
+                    data,
+                    env,
+                })
+            }
+            (None, Some(_)) => warn!("missing window for id {:?}", window_id),
+            (Some(_), None) => warn!("missing state for window id {:?}", window_id),
+            (None, None) => warn!("unknown window {:?}", window_id),
+        }
+        None
+    }
+}
+
+impl<'a, T: Data> SingleWindowState<'a, T> {
+    fn paint(&mut self, piet: &mut Piet, ctx: &mut dyn WinCtx) -> bool {
+        let request_anim = self.send_anim_frame(ctx);
+        self.send_layout(piet);
+        piet.clear(BACKGROUND_COLOR);
+        self.send_paint(piet);
+        request_anim
+    }
+
+    fn send_anim_frame(&mut self, ctx: &mut dyn WinCtx) -> bool {
```

Comment:
Not wild about the "send" naming here as it seems to imply async or deferred operation, but I do see why you want a different name than "do" as below.